### PR TITLE
FCS namespace revamp

### PR DESCRIPTION
**Pull request by dsyme**
This a revamp of the namespace and naming of FCS following on from a design review with @TIHan @jonsequitur @cartermp @KevinRansom

Namespaces are:

    FSharp.Compiler.IO  // FileSystem
    FSharp.Compiler.CodeAnalysis // FSharpChecker, FSharpCheckFileResults, FSharpChecProjectResults and friends
    FSharp.Compiler.Diagnostics     // FSharpDiagnostic and friends
    FSharp.Compiler.EditorServices // Misc functionality for editors, e.g. interface stub generation
    FSharp.Compiler.Interactive.Shell // F# Interactive
    FSharp.Compiler.Symbols           // FSharpEntity etc
    FSharp.Compiler.Syntax           // Types and functionality for syntax tree
    FSharp.Compiler.Text           // ISourceFile, Range, TaggedText and other things
    FSharp.Compiler.Tokenization      // FSharpLineTokenizer etc.

Changes are documented in [release-notes.md](https://github.com/dsyme/fsharp/blob/fcs6/release-notes.md#fsharp-compiler-service-4000)

Other internal cleanup changes:

* dead code removed in `vsintegration/src/FSharp.LanguageService/ProjectSitesAndFiles.fs`

TODO in later PRs:

* [ ] Switch to `ImmutableArray` instead of `IList` through the Symbols API 

* [ ] Add an immutable, extensible, binary-compati-maintainable API to give binary-compatible access to the syntax tree

### CODE REVIEW

File: release-notes.md [hunk 1]
```diff
@@ -50,16 +50,16 @@ This is a big update to FCS. There are significant trimmings and renamings of th
 Renamings:
 
 ```diff
--type FSharp.Compiler.AbstractIL.Internal.Library.IFileSystem
-+type FSharp.Compiler.SourceCodeServices.IFileSystem
+-type FSharp.Compiler.AbstractIL.Library.IFileSystem
```
Comment (cartermp) [inline release-notes.md:53]: This is only correct if we plan on getting this in for FCS 39, which to my knowledge isn't the case? At least not that I'm aware of

Comment (dsyme) [reply release-notes.md:53]: I don't mind - just let me know the schedule.  

(TBH I'm ok with delaying 39 until this is in, even if it's not 100% complete it's still better to get the foundational renames into the ecosystem.  But I don't know if there are people desperate to get 39 or not)

**Comment by dsyme** [comment]
Aside: It did occur to me that perhaps the naming should be as follows to match the DLL, though I don't really mind

```
FSharp.Compiler.Service.IO  // FileSystem
FSharp.Compiler.Service.CodeAnalysis // FSharpChecker, FSharpCheckFileResults, FSharpChecProjectResults and friends
FSharp.Compiler.Service.Diagnostics     // FSharpDiagnostic and friends
FSharp.Compiler.Service.EditorServices // Misc functionality for editors, e.g. interface stub generation
FSharp.Compiler.Service.Syntax           // Types and functionality for syntax tree
FSharp.Compiler.Interactive.Shell // F# Interactive
```

**Comment by NinoFloris** [comment]
`FSharp.CompilerServices` maybe?

File: release-notes.md (hunk 1 above)
Comment (cartermp) [reply release-notes.md:53]: We need to, at a minimum, release FCS 39 with VS 16.9. VS 16.9 bits are locked down for bug fixes only. So I highly doubt we'd incorporate this right now since it's helpful from a coherency standpoint - until we have a shipping cadence for FCS properly established - to have bits map to one another cleanly.

Comment (cartermp) [reply release-notes.md:53]: FCS 39 would come with a bunch of bug fixes and improvements already, so folks definitely would want it

Comment (baronfel) [reply release-notes.md:53]: Yep, eagerly anticipating a drop of 39 (since the nightlies aren't usable for packaging reasons :) ).

Comment (dsyme) [reply release-notes.md:53]: OK great thank you.  So I think that means we can integrate this to master and not into 16.9 release branch?  I'll make sure the release notes are separate under 40.0

File: src/fsharp/TextLayoutRender.fs [hunk 2]
```diff
@@ -167,8 +167,7 @@ module LayoutRender =
 
     let emitL f layout = renderL (taggedTextListR f) layout |> ignore
 
-    let toImmutableArray layout =
```
Comment (cartermp) [inline src/fsharp/TextLayoutRender.fs]: why this change?

Comment (dsyme) [reply src/fsharp/TextLayoutRender.fs]: ImmutableArray is not in the default reference set for F# Interactive (i.e. the default reference set for the F# scripting model) and in my testing this is causing problems when using the API (though I could just reference it in my test scripts)

We can still use ImmutableArray but I just wanted to do it all-at-once intentionally (replacing all uses of `IList<_>` by `ImmutableArray<_>` for example) and also remembering to fix this problem.

I'll make a note of this at the top of this PR

**Comment by dsyme** [comment]
> FSharp.CompilerServices maybe?

I quite like this name.

```
FSharp.CompilerServices.IO  // FileSystem
FSharp.CompilerServices.CodeAnalysis // FSharpChecker, FSharpCheckFileResults, FSharpChecProjectResults and friends
FSharp.CompilerServices.Diagnostics     // FSharpDiagnostic and friends
FSharp.CompilerServices.EditorServices // Misc functionality for editors, e.g. interface stub generation
FSharp.CompilerServices.Interactive.Shell // F# Interactive
FSharp.CompilerServices.Symbols           // FSharpEntity etc
FSharp.CompilerServices.Syntax           // Types and functionality for syntax tree
FSharp.CompilerServices.Text           // ISourceFile, Range, TaggedText and other things
FSharp.CompilerServices.Tokenization      // FSharpLineTokenizer etc.
```

**Comment by dsyme** [comment]
This PR is ready (once green).   It's large but nearly entirely routine. 

Would appreciate getting it merged fairly rapidly to minimise maintenance and get it integrated against all other ongoing work.

File: src/fsharp/SyntaxTree.fsi [hunk 3]
```diff
@@ -0,0 +1,2110 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace rec FSharp.Compiler.Syntax
+
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.Text
+
+/// Represents an identifier in F# code
```
Comment (auduchinok) [inline src/fsharp/SyntaxTree.fsi:8]: Could xml docs be removed from the fs file due to addition of the fsi one, so for each declaration there's a single place where the xml doc is?

Comment (dsyme) [reply src/fsharp/SyntaxTree.fsi:8]: > Could xml docs be removed from the fs file due to addition of the fsi one, so for each declaration there's a single place where the xml doc is?

We have no hard and fast rule in the compiler - in some places we duplicate (e.g. NameResolution.fs) and in others we don't.

I think I'll leave it as is for now.

Comment (dsyme) [reply src/fsharp/SyntaxTree.fsi:8]: I thought about it and agree with you - I removed all the comments from SyntaxTree.fs

**Review by KevinRansom — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Comment by dsyme** [comment]
Thank you for the quick merge!!!
